### DISCUSSION TITLE
Is it possible to wrap a call to reportIssue and still have the runtime warning show on the correct line?

### DISCUSSION THREAD

**Discussion by tfe** [pointfreeco/swift-issue-reporting#109] [Q&A]
I was testing out this library in my project and for [reasons] wanted to wrap calls to it in my own method. I thought by passing in the parameters for `fileID`, `filePath`, etc. that would be sufficient to preserve the original source location, however in practice the runtime warning flag is still showing up on the line where I call `IssueReporting.reportIssue(message, fileID: fileID, filePath: filePath, line: line, column: column)`, which is always in my wrapper function and not terribly useful.

Has anyone successfully done something like this?

Here's what I'm actually trying to do for context... maybe there's a better way to accomplish my root goal that doesn't involve this wrapper?

```swift
/// Wrapper around function of same name from `IssueReporting` that supports
/// additional context data.
///
/// The main reason to do this is because most issues we want to report involve
/// additional context data, which we want to both log and send to an issue
/// monitoring service.
///
/// We accomplish it by partitioning `IssueReporters.current` into custom and
/// built-in ones. We call the custom context-supporting method on our custom ones
/// and then call the regular method with just the built-in reporters.
@_transparent
public func reportIssue(
  _ message: @autoclosure () -> String,
  context: [String: Any],
  fileID: StaticString = #fileID,
  filePath: StaticString = #filePath,
  line: UInt = #line,
  column: UInt = #column
) {
    let message = message()

    /// Loop through current reporters, look for our custom types, and call them with the extra context.
    let nonCustomIssueReporters: [any IssueReporter] = IssueReporters.current.compactMap { issueReporter in
        switch issueReporter {
        case let issueMonitoringReporter as IssueMonitoringReporter:
            issueMonitoringReporter.sendToService(message: message, context: context)
            return nil
        case let contextLoggingReporter as ContextLoggingReporter:
            contextLoggingReporter.logToConsole(message: message, context: context)
            return nil
        default:
            return issueReporter
        }
    }

    /// Send up the chain, with a set of reporters excluding the ones we already called.
    withIssueReporters(nonCustomIssueReporters) {
        IssueReporting.reportIssue(message, fileID: fileID, filePath: filePath, line: line, column: column)
        // ^^^^^^^ runtime warning flags this line ^^^^^^^ 
    }
}
```

**Comment by mbrandonw** [comment]
Hi @tfe, I think in order for this to work our `withIssueReporters` needs to be marked as `@_transparent`. Basically every layer down to `reportIssue` needs to be `@_transparent` in order for the warning to show on the correct line. Want to try forking the repo to see if that works, and if so PR it?

**Reply by tfe** [reply]
Ah, that would explain it, thank you! I tried the suggestion but the compiler is not happy making that function `@_transparent` because of the reference to `$_current` which is private.

I appreciate the reply! Let me know if you can think of any other workarounds.

**Reply by stephencelis** [reply]
@tfe If you make it public can you confirm that it fixes the issue for you? If so we can look into using `@usableFromInline` to traffic access through a computed property.

**Reply by tfe** [reply]
I did try that but there's a whole dependency tree of errors, i.e. making `_current` is an error because `LockIsolated` is an internal type. Trying to make that public causes an error because it is `@usableFromInline` and that can "only be applied to internal or package declarations."

I'm happy to try specific things and report the results back, but I'm way out of my depth here so it's hard for me to understand these errors or see how they might be fixed or worked around.

**Reply by mbrandonw** [reply]
Hi @tfe, I looked into this a bit more and unfortunately I don't think it's actually possible since ultimately `TaskLocal`'s `withValue` is not `@_transparent`, and so that is going to prevent the runtime warning from showing on the correct line. 

Luckily the stack trace of the warning ultimately points to the offending line, but I don't see a clear way to make the warning show up on the line calling your `reportIssue`. It would be really nice if these runtime warnings were officially supported by Apple/Xcode!

**Reply by stephencelis** [reply]
@tfe We also looked over your helper again and think that you could get away with not using `withIssueReporters` at all. There is a version of `reportIssue` that takes an `any Error` as its first parameter, and so your helper could traffic the context through to your underlying reporters using your own error type.

```swift
struct IssueContext: Error {
  var params: [String: Any]
}

struct MyIssueReporter: IssueReporter {
  func reportIssue(
    _ error: any Error, …
  ) {
    guard let error = error as? IssueContext else { fatalError() }
    // Do something with error.params
  }
}
```
